### PR TITLE
use ensurepip instead of get pip.py

### LINKED ISSUE #10: Doesn't work with Yarn PnP
# Error

```
No python_modules directory; installing pip locally if needed.
C:\Python311\python.exe: can't find '__main__' module in 'C:\\Git\\my-package\\.yarn\\cache\\nopy-npm-0.2.8-84f8ff545f-37f4f2bda3.zip\\node_modules\\nopy\\src\\get-pip.py'
Error: Exited with code 1.
    at ChildProcess.<anonymous> (C:\Git\Youtube-Volume-Scroll\.yarn\cache\nopy-npm-0.2.8-84f8ff545f-37f4f2bda3.zip\node_modules\nopy\src\api.js:278:20)
    at ChildProcess.emit (node:events:512:28)
    at maybeClose (node:internal/child_process:1098:16)
    at ChildProcess._handle.onexit (node:internal/child_process:304:5)
```

# Cause

Trying to access a file inside the node_modules folder, which is inside a .zip file if using PnP

https://github.com/alastairpatrick/nopy/blob/06e034b18801541bb4430fb584599a692c1c9ef8/src/npip.js#L17

# Solution

```python
return spawnPython(["-m", "ensurepip", "--upgrade", "--user"], {
```

I've opened a PR (#11) to apply the suggested solution

### PR DESCRIPTION
**Pull request by Araxeus**
fix #10 (Not working with Yarn PnP)

New `installPip` logic:

After every step, return if it succeeds

1. run `pip --version` – fails if `pip` isn't installed
2. run `ensurepip --upgrade --user` – fails if `ensurepip` isn't installed ([Debian / Ubuntu](https://gist.github.com/tiran/2dec9e03c6f901814f6d1e8dad09528e#ensurepip-is-missing))
3. run `./get-pip.py --user --quiet` – fails if using Yarn PnP or pnpm
4. fetch and stream to a temp file `https://bootstrap.pypa.io/get-pip.py`, then run it
  (As a last ditch, this does work with PnP since the saved file is outside the archive)

> Notes:
> * reworked `installPip` to async/await because I got into some insane callback hell
> * I tested the changes on my Windows machine and on an Ubuntu github runner
> * Pretty sure that 4. happen only if Running PnP on Ubuntu with pip not installed... not very likely

### CODE REVIEW

File: src/npip.js [hunk 1]
```diff
@@ -5,27 +5,71 @@
...
+    return;
+  } catch (e) { }
+
+  // run get-pip.py.
+  try {
+    await spawnPython([path.join(__dirname, "get-pip.py"), "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch (error) {
+    console.error("Failed to install pip locally.\n" + JSON.stringify(error));
```
Comment (alastairpatrick) [inline src/npip.js]: I think delete this line since it still has the last resort download to try.

File: src/npip.js [hunk 2]
```diff
@@ -5,27 +5,71 @@
...
+
+  // Check if pip is already installed.
+  try {
+    await spawnPython(["-m", "pip", "-V"], opt);
+    return;
+  } catch (e) { }
+
+  // Try to install via ensurepip.
+  try {
+    await spawnPython(["-m", "ensurepip", "--upgrade", "--user"], opt);
```
Comment (alastairpatrick) [inline src/npip.js]: The goal is only to install pip in the virtual environment if it is not already installed, not to upgrade pip if it is already installed. So I think remove the "--upgrade" option but leave the others.

File: src/npip.js [hunk 3]
```diff
@@ -5,27 +5,71 @@
...
+  try {
+    await spawnPython([path.join(__dirname, "get-pip.py"), "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch (error) {
+    console.error("Failed to install pip locally.\n" + JSON.stringify(error));
+  }
+
+  // Last resort: download pip and run it.
+  const tmpFile = path.join(pkg.dir, "get-pip.py");
```
Comment (alastairpatrick) [inline src/npip.js]: I think get-pip.py should be downloaded to a temporary directory rather than the package directory. Not _the_ temporary directory, because of potential [security issues](https://en.wikipedia.org/wiki/Symlink_race). I think this would do:
```js
await fs.mkdtemp(path.join(os.tmpdir(), 'npip-'));
```
And delete the directory when finished.

File: src/npip.js [hunk 4]
```diff
@@ -5,27 +5,71 @@
...
+      if (!res.ok)
+        throw new Error("Failed to download get-pip.py: " + res.statusText);
+      return res.body.pipe(fs.createWriteStream(tmpFile));
     });
-  });
-}
+    // run get-pip.py
+    await spawnPython([tmpFile, "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    fs.unlinkSync(tmpFile);
```
Comment (alastairpatrick) [inline src/npip.js]: This could be async

File: src/npip.js [hunk 5]
```diff
@@ -5,27 +5,71 @@
...
+    return;
+  } catch (error) {
+    fs.unlinkSync(tmpFile);
```
Comment (alastairpatrick) [inline src/npip.js]: This could be async

**Review by alastairpatrick — COMMENTED** [review]
I think this is generally good. A few things inline.

**Comment by Araxeus** [comment]
All comments have been fixed + some other stuff like `catch { }` instead of `catch (e) { }` and `try catch finally` instead of repeating `unlink`

File: src/npip.js [hunk 6]
```diff
@@ -63,11 +63,11 @@ const installPip = async (pkg) => {
     // run get-pip.py
     await spawnPython([tmpFile, "--user", "--quiet"], opt);
     console.log("Successfully completed pip check.");
-    fs.unlinkSync(tmpFile);
     return;
   } catch (error) {
-    fs.unlinkSync(tmpFile);
     console.error("Failed to install pip locally.\n" + JSON.stringify(error));
```
Comment (Araxeus) [inline src/npip.js:68]: Should this maybe be `throw` instead of `console.error`? Since if we got to this step and it failed, things are looking pretty doomed.

File: src/npip.js [hunk 7]
```diff
@@ -63,11 +63,11 @@ const installPip = async (pkg) => {
     // run get-pip.py
     await spawnPython([tmpFile, "--user", "--quiet"], opt);
     console.log("Successfully completed pip check.");
-    fs.unlinkSync(tmpFile);
     return;
   } catch (error) {
-    fs.unlinkSync(tmpFile);
-    console.error("Failed to install pip locally.\n" + JSON.stringify(error));
+    console.error("Failed to install pip locally.\n" + JSON.stringify(error, null, 2));
```
Comment (Araxeus) [inline src/npip.js]: should this be `throw` instead? since if we got to this point, things are pretty doomed?

Comment (alastairpatrick) [reply src/npip.js]: Yeah I think either don't catch the exception in the first place or keep the catch block but throw the exception after outputting the error message.

File: src/npip.js [hunk 8]
```diff
@@ -3,29 +3,73 @@
...
-  });
-}
+    // run get-pip.py
+    await spawnPython([tmpFile, "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch (error) {
+    console.error("Failed to install pip locally.\n" + JSON.stringify(error, null, 2));
+  } finally {
+    await fs.promises.unlink(tmpFolder);
```
Comment (alastairpatrick) [inline src/npip.js]: I think unlink() will not work on directories. Maybe:
```js
await fs.promises.rmdir(tmpFolder, { recursive: true });
```
But that runs the risk of deleting files that were not intended. Below would be safer but would fail if the directory was created but the file does not exist:
```js
await fs.promises.unlink(tmpFile);
await fs.promises.rmdir(tmpFolder);
```
Maybe ignore unlink failing if the reason is ENOENT.

Comment (Araxeus) [reply src/npip.js]: woupsie my bad

File: src/npip.js (hunk 7 above)
Comment (Araxeus) [reply src/npip.js]: Opted for removing the catch block

File: src/npip.js (hunk 8 above)
Comment (Araxeus) [reply src/npip.js]: >  `fs.rmdir()` But that runs the risk of deleting files that were not intended

How so?

we just made that folder, why would there be anything else in there?

Comment (alastairpatrick) [reply src/npip.js]: Like if someone was hacking on the npip code and accidentally made tmpFolder point at /tmp, it would try to delete /tmp recursively. If you've run it and your /tmp directory survived then recursive version is fine :)

Comment (Araxeus) [reply src/npip.js]: It should be fine

```js
  const tmpFolder = await fs.promises.mkdtemp(path.join(os.tmpdir(), 'npip-'));
```
this will always be a random temp folder or at the very least an error

**Comment by alastairpatrick** [comment]
I tested this on a Windows machine with node 18.14.2, npm 9.5.0 and no global pip installation. I commented out various blocks of code and each time deleted the python_modules directory to make it perform the various methods of installing pip. All worked except the last resort download. I initially got the error below. 

```
D:\Temp\myproject>npm run -s npip -- install
No python_modules directory; installing pip locally if needed.
(node:35280) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
(Use `node --trace-deprecation ...` to show where the warning was created)
Error: ENOTEMPTY: directory not empty, rmdir 'C:\Users\Al\AppData\Local\Temp\npip-RSuIxt'
```

Replacing rmdir() with rm() is easy but I'm confused as to why it's complaining that the directory is not empty when it's supposed to be deleting it recursively.

Next, I commented out the rmdir() line and tried again. I got this error and obviously it didn't install pip:

```
D:\Temp\myproject>npm run -s npip -- install
No python_modules directory; installing pip locally if needed.
TypeError: res.body.pipe is not a function
    at D:\Temp\myproject\node_modules\nopy\src\npip.js:61:23
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async installPip (D:\Temp\myproject\node_modules\nopy\src\npip.js:53:5)
```

**Comment by Araxeus** [comment]
Yikes I used `node-fetch` api, while actually using the native `fetch`

**Comment by Araxeus** [comment]
[`1cc423a` (#11)](https://github.com/alastairpatrick/nopy/pull/11/commits/1cc423a9b9c4cf86b77276a6defccfbfce7ac9d4)
* changed the native `fetch` to `https.get` (doesn't require additional deps)
* use `rm` instead of `rmdir`

I'm really sorry for not testing that last part… feeling really stupid right now
I actually added it at the last second, just for an extreme edge case, then forgot to test it

This time I properly tested all steps. 
once again sorry for the messy PR, and thank you for taking the time to review and test it
